### ISSUE TITLE
New lint suggestion `Error impl Error`

### ISSUE THREAD

**Issue by JonathanWoollett-Light** [rust-lang/rust-clippy#11101]
### What it does

Warns on types which implement `std::error::Error` being named `Error`.

### Advantage

Removes needing to document against using this common pattern.
The reasons against this pattern is that it can reduce readability and requires error types be renamed when imported (which can further reduce readability).

When you have 20 types named `Error` in your project it can become annoying.

### Drawbacks

Adds an additional lint to clippy.

### Example

```rust
#[derive(Debug, thiserror::Error)]
pub enum Error { /* ... */ }
```

Could be written as:

```rust
#[derive(Debug, thiserror::Error)]
pub enum MySpecificError { /* ... */ }
```

```rust
#[derive(Debug)]
pub enum Error { /* ... */}
impl std::error::Error for Error { /* ... */ }
```

Could be written as:

```rust
#[derive(Debug)]
pub enum MySpecificError { /* ... */}
impl std::error::Error for Error { /* ... */ }
```

**Comment by y21** [comment]
This is a pretty common pattern (even in the standard library there's `io::Error` and `fmt::Error`) so imo this should be a restriction lint if it gets added
